File: src/com/announcify/web/gwt/client/Background.java
package com.announcify.web.gwt.client;

import com.google.gwt.core.client.EntryPoint;
import com.google.gwt.user.client.Window;

public class Background implements EntryPoint {

	@Override
	public void onModuleLoad() {
		addBrowserAction(this);
	}


	private void onBrowserActionClicked(String url) {
		// TODO: lang
		Window.open(getHtmlUrl(this) + "?url=" + url, "announcify.web", null);
	}


	private native void addBrowserAction(Background background) /*-{
		chrome.browserAction.onClicked.addListener(function (tab) {
			background.@com.announcify.web.gwt.client.Background::onBrowserActionClicked(Ljava/lang/String;)(tab.url);
		});
	}-*/;
	
	private native String getHtmlUrl(Background background) /*-{
		return chrome.extension.getURL("html/announcify.html");
	}-*/;
}

File: script/background.js
if (!localStorage.openedSettings) {
    window.open(chrome.extension.getURL("html/options.html"));

    localStorage.openedSettings = true;
}

chrome.browserAction.onClicked.addListener(function (tab) {
    chrome.tabs.detectLanguage(tab.id, function (language) {
		chrome.tabs.executeScript(tab.id, {code: "(" + getSelectionAndAnnouncify.toString() + ")('" + language + "');"});
	});
});

chrome.extension.onRequest.addListener(function(request, sender, sendResponse) {
    if (request.type == 'announcify.web') {
        var url = chrome.extension.getURL("html/announcify.web.html") + "?lang=" + request.lang;
        url += '&url=' + request.url + '&title=' + request.title + '&selected=' + request.selected;

        if (request.selected) {
            url += '&text=' + request.text;
        }

        window.open(url, "announcify.web");

        trackPage(request);
    } else {
        trackEvent(request);
    }
});


function getSelectionAndAnnouncify(language) {
	var request = {type: 'announcify.web', selected: false, lang: language};
    request.hostname = window.location.hostname;
    request.url = window.location.href;
    request.title = escape(document.title);

	if (window.getSelection().toString()) {
		request.text = escape(window.getSelection().toString());
	}

    chrome.extension.sendRequest(request);
}

function trackPage(request) {
	_gaq.push(['_trackEvent', 'announcify.web.domain', request.hostname]);
    _gaq.push(['_trackEvent', 'announcify.web.url', request.url]);
    _gaq.push(['_trackEvent', 'announcify.web.selected', request.selected]);
}

function trackEvent(event) {
	_gaq.push(['_trackEvent', 'announcify.web.' + event.name, event.value]);
}
